### PR TITLE
Expose `Q2`

### PR DESCRIPTION
**Pull request by OverLordGoldDragon**
For JTFS use cases with `Q2 > 1`, recommended by @changhongw and @lostanlen

Exposes to `Scattering1D` and all frontend derivatives, since JTFS reuses `Scattering1D`. Updated docstrings everywhere in 1D.

### CODE REVIEW

**Comment by lostanlen** [comment]
Thanks. i think that the API we had in mind with @Janden and others would be to pass a tuple `Q=(Q1, Q2)` instead of passing `Q1` and `Q2` as separate kwargs
At least this is what we did in ScatNet

But i'm also fine with merging it this way and revisiting the API later

**Comment by OverLordGoldDragon** [comment]
@lostanlen Considered it, but this is clearer to document (and talk about). They're distinct parameters, best made explicit.

**Comment by OverLordGoldDragon** [comment]
I'm mainly for explicitness in `TimeFrequencyScattering`, where we make room to describe various use cases; no need for it in `Scattering1D`. Former could pass to latter as a tuple (though, it creates some internal inconsistency with arg handling).

**Comment by lostanlen** [comment]
it's passig on `pip` and failing on `conda`. related to the infamous `torch.fft module is not callable` issue (#723)

merge?

**Comment by OverLordGoldDragon** [comment]
@lostanlen I reproduce #723 on dev but not master, so dev also has refactoring to do relative to master. Should we be recommending `master` over `v0.2.0`, or `dev`, to users? @edouardoyallon 

I'm fine with merging.

**Comment by changhongw** [comment]
> For JTFS use cases with `Q2 > 1`, recommended by @changhongw and @lostanlen
> 
> Exposes to `Scattering1D` and all frontend derivatives, since JTFS reuses `Scattering1D`. Updated docstrings everywhere in 1D.
> 
> ### Todo
> 1. Expose to `Scattering1D`, or only internally?

This depends on the developing team's long-term plan for Kymatio? @lostanlen Passing Q1 and Q2 as a tuple Q=(Q1, Q2) helps the user be aware of the changeability of Q2 and the scattering path. If in the feature, we would like to explore higher order scattering, i.e. to have a Q3 (filters per octave of the third-order time scattering), it could be naturally Q=(Q1, Q2, Q3)? Similarly, we could also have Q_fr = (Q_fr1, Q_fr2). In such cases, a tuple would be better than many separate variables for the argument of `Scattering1D`? 

For the second-order, either works for me as well. The current way introduces fewer changes to `Scattering1D`.

**Comment by OverLordGoldDragon** [comment]
> if in the future ... higher order

Agreed, then tuple's certainly preferable.

I'm not strongly disposed either way, though any higher order for JTFS seems a long shot, so maybe explicit `Q2` just for JTFS.

**Comment by lostanlen** [comment]
i've never implemented third-order JTFS. i imagine it's very intensive computationally. not a priority for Kymatio IMO.
I would prefer something along the lines of
`TimeFrequencyScattering -> Conv2D -> Scattering2D -> Conv2D` ...

So i'm fine with `TimeFrequencyScattering` having exactly two kwargs: `Q` and `Q2`.

Third-order time scattering could be useful for some non-audio applications, but the code is nowhere near ready for that yet. Maybe we won't ever have it (https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)
Let's wait till we have a specific use case before worrying about `Q=(Q1,Q2,Q3)`.

In conclusion i agree with @OverLordGoldDragon 's design

waiting on @janden's review and then we can merge

**Comment by janden** [comment]
I don't quite follow the argument for separate kwargs. Seems like the tuple approach would be cleaner here (after all, it's `Q`, not `Q1`). For internals, we could separate, but for API I think a tuple is more appropriate.

**Comment by OverLordGoldDragon** [comment]
@janden Current plan is `Q`-only (`Q1`) or tuple for `Scattering1D` and dedicated for `TimeFrequencyScattering`. I'm for making `Q2` explicit so it's more obvious the option exists, and since it's clearer to document and refer to.

Let me know of the agreement, I'll implement - either will work for me.

**Comment by janden** [comment]
> Current plan is `Q`-only (`Q1`) or tuple for `Scattering1D`

I'm not following exactly. From the code, I'm not seeing a tuple, but two arguments: `Q` and `Q2`. Again, I think it would be simpler to have a tuple argument here.

> dedicated for `TimeFrequencyScattering`

Dedicated `Q2` or dedicated `Q_fr`? I think the latter would make more sense here in terms of compatibility with TS.

**Comment by OverLordGoldDragon** [comment]
> I'm not seeing a tuple

Right, I've yet to change it. Then I'll make `Q` a tuple in `Scattering1D`.

> Dedicated `Q2` or dedicated `Q_fr`? I think the latter would make more sense here in terms of compatibility with TS.

Dedicated both.

So then: tuple for both, or dedicate `Q2` in JTFS?

**Comment by janden** [comment]
I'd say tuple `Q` for both to make things compatible, then another argument `Q_fr` to control the frequency Q in JTFS.

**Comment by OverLordGoldDragon** [comment]
@janden Adjusted.

**Comment by lostanlen** [comment]
i need this feature urgently. considering that the test failure is unrelated (#723), i'd be in favor of merging (to `dev`)

File: kymatio/scattering1d/frontend/base_frontend.py [hunk 1]
```diff
@@ -283,8 +287,9 @@ def output_size(self, detail=False):
             The maximum log-scale of the scattering transform. In other words,
             the maximum scale is given by `2 ** J`.
         {param_shape}Q : int
-            The number of first-order wavelets per octave (second-order
```
Comment (janden) [inline kymatio/scattering1d/frontend/base_frontend.py:286]: Needs to be fixed to align with docstring above.

**Comment by OverLordGoldDragon** [comment]
@lostanlen Use the JTFS branch?

**Comment by lostanlen** [comment]
@OverLordGoldDragon good idea, thanks

File: kymatio/scattering1d/frontend/keras_frontend.py [hunk 2]
```diff
@@ -7,15 +7,16 @@
 
 
 class ScatteringKeras1D(ScatteringKeras, ScatteringBase1D):
-    def __init__(self, J, Q=1, max_order=2, oversampling=0):
```
Comment (janden) [inline kymatio/scattering1d/frontend/keras_frontend.py:10]: Shouldn't this just be `Q` to allow tuples to pass through? Same with the other frontends.

**Review by janden — CHANGES_REQUESTED** [review]
Seems like a few `Q2`s slipped through.

File: kymatio/scattering1d/frontend/base_frontend.py (hunk 1 above)
Comment (OverLordGoldDragon) [reply kymatio/scattering1d/frontend/base_frontend.py:286]: Same align as original, no?

File: kymatio/scattering1d/frontend/keras_frontend.py (hunk 2 above)
Comment (OverLordGoldDragon) [reply kymatio/scattering1d/frontend/keras_frontend.py:10]: Missed frontends

**Comment by OverLordGoldDragon** [comment]
JTFS is ahead of most these PRs, I no longer replicate every change. Adjusted this one.

File: kymatio/scattering1d/frontend/base_frontend.py (hunk 1 above)
Comment (janden) [reply kymatio/scattering1d/frontend/base_frontend.py:286]: This one talks about Q, Q2, while the above docstring talks about tuples.

**Comment by janden** [comment]
> JTFS is ahead of most these PRs, I no longer replicate every change. Adjusted this one.

So are these PRs meant to be merged or just staging for JTFS?

File: kymatio/scattering1d/frontend/base_frontend.py (hunk 1 above)
Comment (OverLordGoldDragon) [reply kymatio/scattering1d/frontend/base_frontend.py:286]: ah _that_ 'align'

**Comment by OverLordGoldDragon** [comment]
Latter; former involves figuring out a merge order and successively fixing conflicts - redundant effort. I can guarantee the PRs accurately reflect what'll be merged with JTFS. (but if anyone's willing to do it I won't object)

I suggest instead keeping the PRs open and referring to them while reviewing JTFS

**Comment by janden** [comment]
Ok.

First, if a PR is not meant to be reviewed and merged, please mark it as a draft and state as much. Furthermore, if there's a discussion about a particular aspect of JTFS that needs to happen, it's better to take that through issues or the discussion board or discord.

Second, I don't think it's a good idea to merge JTFS in one mega-PR. It currently amounts to 4000 additional lines and it will be unmanageable to review it properly. If merging these smaller PRs is equivalent, that's preferable.

**Comment by OverLordGoldDragon** [comment]
So I misunderstood how concurrent merges work.

Yes, I favor merging these PRs first, and can recommend a merge order. #673 should come first (but maybe after #748).

> issues or the discussion board or discord.

I find it better to keep things all in one place. Else there's thread duplication and need to check many sources for one thing.

There's already plenty in discussions and discord regarding JTFS, the PRs are in "almost ready" state at worst.

**Review by MuawizChaudhary — COMMENTED** [review]
LGTM
